src/java.desktop/macosx/native/libawt_lwawt/awt/a11y/OutlineAccessibility.m

```diff
@@ -55,4 +55,88 @@ - (NSString *)accessibilityLabel
     return [[super accessibilityLabel] isEqualToString:@"list"] ? @"tree" : [super accessibilityLabel];
 }
 
+- (nullable NSArray<id<NSAccessibilityRow>> *)accessibilityRows
+{
+    return [self accessibilityChildren];
+}
+
+- (nullable NSArray<id<NSAccessibilityRow>> *)accessibilitySelectedRows
+{
+    return [self accessibilitySelectedChildren];
+}
+
+- (nullable  NSArray<id<NSAccessibilityRow>> *)accessibilityChildren
+{
+    if (![self isCacheValid]) {
+        NSArray *t = [super accessibilityChildren];
+        if (t != NULL) {
+            rowCache = [[NSMutableArray arrayWithArray:t] retain];
+        } else {
+            rowCache = NULL;
+        }
+        rowCacheValid = YES;
+    }
+    return rowCache;
+}
+
+- (nullable NSArray<id<NSAccessibilityRow>> *)accessibilitySelectedChildren
+{
+    if (!selectedRowCacheValid) {
+        NSArray *t = [super accessibilitySelectedChildren];
+        if (t != NULL) {
+            selectedRowCache = [[NSMutableArray arrayWithArray:t] retain];
+        } else {
+            selectedRowCache = NULL;
+        }
+        selectedRowCacheValid = YES;
+    }
+    return selectedRowCache;
+}
+
+- (BOOL)isCacheValid
+{
+    if (rowCacheValid && [[self parent] respondsToSelector:NSSelectorFromString(@"isCacheValid")]) {
```

Comment:
When `isCacheValid` is invoked for the first time then `rowCacheValid` is _FALSE_ and we return _FALSE_ value from this method and then the `accessibilityChildren` method from super class is invoked to get all children and store them in an array(`rowCache`). Subsequently `rowCacheValid` will be set to _TRUE_.

As per my understanding `[[self parent] respondsToSelector:NSSelectorFromString(@"isCacheValid")]` condition will be checked only if `rowCacheValid` is true and once we have a valid array of children then what is the point of checking `isCacheValid` method in parent ?

Reply:
In multilevel trees when someone in the parent's path is collapsed the treeCollapsed event will be delivered to that node. We can either invalidate all leaf nodes caches (which is costly) or mark that exact node's cache as invalid and when the expand will happen when a11y cursor will try to position inside one of the childrens it will check that parent was invalidated and will regenerate its own cache.

Reply:
Ok.
Is it possible to add a test to verify this behavior?

Reply:
Ok. I was thinking of something like when nodes are collapsed or expanded then can we log some output like Children are valid or invalid.
Otherwise I have tested the fix with swingset2 demo and the test mentioned in [JDK-8320664](https://bugs.openjdk.org/browse/JDK-8320664) which behaves fine with the fix.